### PR TITLE
Thchs30 mfa

### PR DESCRIPTION
**Pull request by yt605155624**
### PR types

New features 
### PR changes

Models | APIs
### Describe

add MFA example for THCHS30
here is the result for THCHS30 A11_0.wav
phone -> phone:
phone2phone

syllable - > phone:
syable2phone

text -> phone:
text2phone

### CODE REVIEW

File: examples/thchs30/a0/run.sh [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+fi
+
+# reorganize dataset for MFA
+if [ ! -d $EXP_DIR/thchs30_corpus ]; then
+    echo "reorganizing thchs30 corpus..."
+    python local/recorganize_thchs30.py --root-dir=./data --output-dir=$EXP_DIR/thchs30_corpus --script-type=$LEXICON_NAME
+    echo "reorganization done."
+fi
+# MFA is in tools
+export PATH="${MAIN_ROOT}/tools/montreal-forced-aligner/bin"
```
Comment (zh794390558) [inline examples/thchs30/a0/run.sh]: 加到path.sh中

File: examples/thchs30/a0/path.sh [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+# Use UTF-8 in Python to avoid UnicodeDecodeError when LC_ALL=C
+export PYTHONIOENCODING=UTF-8 
+export PYTHONPATH=${MAIN_ROOT}:${PYTHONPATH}
+
+export LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:/usr/local/lib/
+
+
+MODEL=deepspeech2
+export BIN_DIR=${MAIN_ROOT}/deepspeech/exps/${MODEL}/bin
```
Comment (zh794390558) [inline examples/thchs30/a0/path.sh]: 可以删除13-14

File: examples/thchs30/a0/run.sh [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+source ${MAIN_ROOT}/utils/parse_options.sh || exit 1;
+
+# download dataset、unzip and generate manifest 
+if [ ${stage} -le 0 ] && [ ${stop_stage} -ge 0 ]; then
+    # prepare data
+    bash ./local/data.sh || exit -1
+fi
+
+# reorganize dataset for MFA
+if [ ! -d $EXP_DIR/thchs30_corpus ]; then
```
Comment (zh794390558) [inline examples/thchs30/a0/run.sh]: 可以放到data.sh里完成

File: examples/thchs30/a0/local/recorganize_thchs30.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        # find thchs30_pinyin2phone in dir of this py file
+        py_dir_path = os.path.split(os.path.realpath(__file__))[0]
+        py_dir_path = Path(py_dir_path).expanduser()
+        lexicon_path = py_dir_path / "thchs30_pinyin2phone"
+    else:
+        # script_type == 'text'
+        # find thchs30_cn2phone in dir of this py file
+        py_dir_path = os.path.split(os.path.realpath(__file__))[0]
+        py_dir_path = Path(py_dir_path).expanduser()
+        lexicon_path = py_dir_path / "thchs30_cn2phone"
```
Comment (zh794390558) [inline examples/thchs30/a0/local/recorganize_thchs30.py]: thchs30_cn2phone -> thchs30_word2phone

File: examples/thchs30/a0/local/recorganize_thchs30.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            os.path.dirname(wav_path) + os.path.sep + "..")
+        grader_father = Path(grader_father).expanduser()
+        lexicon_name = "lexicon.txt"
+        lexicon_father_dir = "lm_phone"
+        lexicon_path = grader_father / lexicon_father_dir / lexicon_name
+    elif script_type == 'syllable':
+        # find thchs30_pinyin2phone in dir of this py file
+        py_dir_path = os.path.split(os.path.realpath(__file__))[0]
+        py_dir_path = Path(py_dir_path).expanduser()
+        lexicon_path = py_dir_path / "thchs30_pinyin2phone"
```
Comment (zh794390558) [inline examples/thchs30/a0/local/recorganize_thchs30.py]: thchs30_syllable2phone

File: examples/thchs30/a0/local/recorganize_thchs30.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    parser.add_argument(
+        "--output-dir",
+        type=str,
+        help="path to save outputs(audio and transcriptions)")
+
+    parser.add_argument(
+        "--script-type",
+        type=str,
+        default="phone",
+        help="type of lab (text'/'syllable'/'phone')")
```
Comment (zh794390558) [inline examples/thchs30/a0/local/recorganize_thchs30.py]: text -> word

File: examples/thchs30/a0/run.sh [hunk 7]
```diff
@@ -0,0 +1,40 @@
+#!/bin/bash
+set -e
+source path.sh
+stage=0
+stop_stage=100
+EXP_DIR=exp
+# LEXICON_NAME in {'phone', 'syllable', 'text'}
```
Comment (zh794390558) [inline examples/thchs30/a0/run.sh]: text -> word

File: examples/thchs30/a0/run.sh [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+LEXICON_NAME='phone'
+# get machine's cpu core number
+NUM_JOBS=`grep 'processor' /proc/cpuinfo | sort -u | wc -l`
```
Comment (zh794390558) [inline examples/thchs30/a0/run.sh]: 替换成 `nproc`

Comment (yt605155624) [reply examples/thchs30/a0/run.sh]: all done in https://github.com/PaddlePaddle/DeepSpeech/pull/698/commits/00017301c6c739eed2328d5c42b459ec95d1b78f
rename lexicons to data/dict/{phone, word, syllabl}.lexicon
dump manifest in {MAIN_PATH}/utils/dump_manifest.py
put all data processing scripts in data.sh
put training data in data/ results in exp/ and scripts in local/
what's more :
if len(cn_list) == len(phone_list) / 2 must be true in THCHS30's origin word -- phone lexicons (except for some error lines), a Chinese character is consist of two phones in THCHS30's standard

File: examples/thchs30/README.md [hunk 9]
```diff
@@ -1,2 +1,42 @@
...
+下载 MFA 的可执行包（也会同时下载本项目所需的其他工具）
+执行如下命令：
+```
+cd a0
+./run.sh
+```
+应用程序会自动下载 THCHS-30数据集，处理成 MFA 所需的文件格式并开始训练，您可以修改 `run.sh` 中的参数 `LEXICON_NAME` 来决定您需要强制对齐的级别（word、syllable 和 phone）
+## MFA 所使用的字典
+---
+MFA 字典的格式请参考: [MFA 官方文档 Dictionary format ](https://montreal-forced-aligner.readthedocs.io/en/latest/dictionary.html)
```
Comment (zh794390558) [inline examples/thchs30/README.md:30]: 可以关注下MFA的字典生成工具和过程，多音字的probs和不一定等于1.0

File: examples/thchs30/a0/local/data.sh [hunk 10]
```diff
@@ -22,5 +23,25 @@ if [ ${stage} -le -1 ] && [ ${stop_stage} -ge -1 ]; then
     
 fi
 
+# dump manifest to data/
```
Comment (zh794390558) [inline examples/thchs30/a0/local/data.sh:26]: 加上stage控制步骤。

File: examples/thchs30/a0/local/gen_word2phone.py [hunk 11]
```diff
@@ -53,9 +53,9 @@ def gen_lexicon(root_dir: Union[str, Path], output_dir: Union[str, Path]):
     root_dir = Path(root_dir).expanduser()
     output_dir = Path(output_dir).expanduser()
     output_dir.mkdir(parents=True, exist_ok=True)
-    file1 = root_dir / "data_thchs30/lm_word/lexicon.txt"
-    file2 = root_dir / "resource/dict/lexicon.txt"
-    write_file = output_dir / "thchs30_cn2phone"
+    file1 = root_dir / "lm_word_lexicon_1"
```
Comment (zh794390558) [inline examples/thchs30/a0/local/gen_word2phone.py:56]: 通过参数传入，写死了不好看代码。

File: examples/thchs30/a0/run.sh [hunk 12]
```diff
@@ -4,31 +4,23 @@ source path.sh
...
-# reorganize dataset for MFA
-if [ ! -d $EXP_DIR/thchs30_corpus ]; then
-    echo "reorganizing thchs30 corpus..."
-    python local/recorganize_thchs30.py --root-dir=./data --output-dir=$EXP_DIR/thchs30_corpus --script-type=$LEXICON_NAME
-    echo "reorganization done."
-fi
-# MFA is in tools
-export PATH="${MAIN_ROOT}/tools/montreal-forced-aligner/bin"
-
+# run MFA
```
Comment (zh794390558) [inline examples/thchs30/a0/run.sh:20]: 用stage隔离不同的步骤。

File: examples/thchs30/a0/local/data.sh [hunk 13]
```diff
@@ -22,5 +23,25 @@ if [ ${stage} -le -1 ] && [ ${stop_stage} -ge -1 ]; then
...
+
+# copy files to data/dict to gen word.lexicon
+cp  ${TARGET_DIR}/thchs30/data_thchs30/lm_word/lexicon.txt data/dict/lm_word_lexicon_1
+cp  ${TARGET_DIR}/thchs30/resource/dict/lexicon.txt data/dict/lm_word_lexicon_2
+
+# copy phone.lexicon to data/dict
+cp  ${TARGET_DIR}/thchs30/data_thchs30/lm_phone/lexicon.txt data/dict/phone.lexicon
+
+# gen word.lexicon
+python local/gen_word2phone.py  --root-dir=data/dict --output-dir=data/dict
```
Comment (zh794390558) [inline examples/thchs30/a0/local/data.sh:37]: 需要清晰的知道输入和输出是什么。

File: examples/thchs30/a0/local/gen_word2phone.py [hunk 14]
```diff
@@ -53,9 +53,9 @@ def gen_lexicon(root_dir: Union[str, Path], output_dir: Union[str, Path]):
...
+    file2 = root_dir / "lm_word_lexicon_2"
+    write_file = output_dir / "word.lexicon"
```
Comment (zh794390558) [inline examples/thchs30/a0/local/gen_word2phone.py:58]: 可以加入文件的输入和输出格式例子注释

**Review by zh794390558 — APPROVED** [review]
LGTM

**Pull request merged**